magma/types/decoupled.py

```diff
@@ -0,0 +1,174 @@
+
+    The base protocol is undirected and is meant to be used with the modifiers:
+    * m.Consumer(m.ReadyValid[T])  # data/valid are inputs, ready is output
+    * m.Producer(m.ReadyValid[T])  # data/valid are outputs, ready is input
+
+    T is the type of data to be wrapped in Ready/Valid
+
+    The actual semantics of ready/valid are enforced via the use of concrete
+    subclasses.
+    """
+    def fired(self):
+        """
+        Returns a bit that is high when ready and valid are both high
+        """
+        raise NotImplementedError
+
+
+class ReadyValidProducerKind(ReadyValidKind):
+    def __getitem__(cls, T: Union[Type, MagmaProtocol]):
+        cls._check_T(T)
+        fields = {"valid": Out(Bit), "data": Out(T), "ready": In(Bit)}
+        return type(f"{cls}[{T}]", (cls, ), fields)
+
+
+class ReadyValidProducer(ReadyValid, metaclass=ReadyValidProducerKind):
+    def fired(self):
+        """
+        Returns a bit that is high when ready and valid are both high
+        """
+        if not self.valid.driven():
+            raise ReadyValidException(
+                "Cannot invoke fired on producer when valid isn't driven"
+            )
+        return self.valid.value() & self.ready
+
+    def enq(self, value, when=True):
+        """
+        Produces the data `value` when `when` is high
+
+        Uses "last connect semantics", which allows multiple `enq` methods to
+        be invoked with different conditions with precedence given to the later
+        invocations.
+        """
+        if when is not True:
+            if not (self.valid.driven() and self.data.driven()):
+                raise ReadyValidException(
+                    "Cannot use enq when without default driver"
+                )
+            curr_valid = self.valid.value()
+            when |= curr_valid
+            self.valid.unwire(curr_valid)
+
+            curr_data = self.data.value()
+            value = mux([curr_data, value], when)
+            self.data.unwire(curr_data)
+        self.valid @= when
+        self.data @= value
+        return value
+
+    def no_enq(self, when=True):
+        """
+        Indicates no enqueue occurs.  Valid is set to false, and bits are
+        connected to 0 (TODO: Should be an unitialized wire (Don't Care))
+        """
+        if when is not True:
+            if not (self.valid.driven() and self.data.driven()):
+                raise ReadyValidException(
+                    "Cannot use no_deq with condition without an existing"
+                    "valid and data driver"
+                )
+            valid = self.valid.value()
+            data = self.data.value()
+            self.valid.unwire(valid)
+            self.data.unwire(data)
+            self.valid @= mux([valid, False], when)
+            self.data @= mux([data, 0], when)
+        else:
+            self.valid @= False
+            self.data @= 0
+
+
+class ReadyValidConsumerKind(ReadyValidKind):
```

Comment:
may want to explicitly disallow qualify on ReadyValidConsumerKind and ReadyValidProducerKind.

Reply:
Tried this, but the internal code uses flip/qualify when dealing with interfaces, so we'd have to change that to give it a private interface that allows it.

Reply:
Hmm not clear to me why qualify ever gets called at the top level (i.e. on the tuple type itself, it makes sense to be called on the fields of tuple).

If that is the case, we could still potentially warn if the top level type is qualified with `In` or `Out` -- this seems unambiguously like something we don't want.

My main concern is that people are so used to doing `x=m.In(...), y=m.Out(...)` that it might be a common thing users do they first time they use a specialized (pre-directed) type like ReadyValidConsumer. It might be good to catch and warn on that earlier than later. Am ok with any reasonable way of doing so, if possible at all.

Reply:
https://github.com/phanrahan/magma/blob/8e038e688776e6ab9f84885dafa0846f449ae680/magma/interface.py#L60

When we're defining circuits with a ready valid port, we need to flip the type to distinguish between the instance versus circuit definition view (internally a port is an input, externally it is an output)

Reply:
(And flip calls qualify)

Reply:
We could allow flip, but not In/Out